tests/unit/s2n_recv_greedy_test.c

```diff
@@ -0,0 +1,518 @@
+        }
+
+        DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+        EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client, server, &io_pair));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server, client));
+
+        struct s2n_recv_wrapper counter = { 0 };
+        EXPECT_OK(s2n_connection_set_counting_read(server, &counter));
+
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+        for (size_t i = 1; i <= sizeof(test_data); i++) {
+            EXPECT_EQUAL(s2n_send(client, test_data, i, &blocked), i);
+        }
+
+        uint8_t buffer[sizeof(test_data)] = { 0 };
+        for (size_t i = 1; i <= sizeof(test_data); i++) {
+            EXPECT_EQUAL(s2n_recv(server, buffer, i, &blocked), i);
+            EXPECT_BYTEARRAY_EQUAL(buffer, test_data, i);
+
+            if (greedy) {
+                /* All our small records combined are smaller than the maximum
+                 * TLS record size, so they should all be buffered immediately.
+                 * Only one read is ever necessary.
+                 */
+                EXPECT_EQUAL(counter.count, 1);
+            } else {
+                /* We call recv twice for every record */
+                EXPECT_EQUAL(counter.count, i * 2);
+            }
+        }
+
+        /* The input buffer size does not change with greedy vs not greedy */
+        EXPECT_EQUAL(server->buffer_in.blob.allocated, buffer_in_size);
+
+        /* If all data is consumed, the input buffer can be released */
+        EXPECT_SUCCESS(s2n_connection_release_buffers(server));
+        EXPECT_EQUAL(server->buffer_in.blob.allocated, 0);
+    }
+
+    /* Test: Read the rest of a partial record */
+    for (size_t i = 0; i < test_record_size; i++) {
+        DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+        DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_SUCCESS(s2n_connection_set_config(server, config));
+
+        EXPECT_SUCCESS(s2n_connection_set_recv_buffering(client, true));
+        EXPECT_SUCCESS(s2n_connection_set_recv_buffering(server, true));
+
+        DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+        EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+        EXPECT_OK(s2n_connections_set_io_stuffer_pair(client, server, &io_pair));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server, client));
+
+        struct s2n_recv_wrapper counter = { 0 };
+        EXPECT_OK(s2n_connection_set_counting_read(server, &counter));
+
+        size_t expected_count = 0;
+
+        /* Test: manually copy some of the record into the read buffer */
+        if (false) {
```

Comment:
this whole branch isn't reachable. is it supposed to be?

Reply:
🤦